### PR TITLE
More documentation on Circom code

### LINKED ISSUE #483: Add more detailed documentation to Semaphore V4 contracts and libraries
**Tasks**:

- [x] Circuits 
- [x] `@semaphore-protocol/contracts`
- [x] Core packages
    - [x] `@semaphore-protocol/identity`
    - [x] `@semaphore-protocol/group`
    - [x] `@semaphore-protocol/proof`
- [x] `@semaphore-protocol/utils`
- [x] `@semphore-protocol/data` https://github.com/semaphore-protocol/semaphore/pull/768
- [x] `@semaphore-protocol/hardhat` https://github.com/semaphore-protocol/semaphore/pull/772
- [x] `@semaphore-protocol/cli` https://github.com/semaphore-protocol/semaphore/pull/771

### PR DESCRIPTION
**Pull request by cedoor**
## Description

This PR adds more documentation to the Semaphore circuits, with the aim of providing an overview of the code used and referring to other resources for further information.

## Related Issue(s)

#483

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [x] I have run `yarn prettier` and `yarn lint` without getting any errors
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes
-   [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/circuits/semaphore.circom [hunk 1]
```diff
@@ -4,22 +4,57 @@ include "babyjub.circom";
...
+// of the Merkle tree, i.e. the Semaphore group. That is, a proof of membership is verified.
+// The third part covers the generation of a nullifier, i.e. the hash of the secret used to
+// derive the public key and the scope of the proof. The nullifier is used to prevent the same
+// proof from being verified twice.
+// The circuit lastly includes the message, which is an arbitrary anonymous value defined by
+// the user, or the hash of that value.
 template Semaphore(MAX_DEPTH) {
+    // Input signals.
+    // The input signals are all private except 'message' and 'scope'.
     signal input secret;
```
Comment (vplasencia) [inline packages/circuits/semaphore.circom:28]: It could be nice to say what is the secret here. What do you think?

**Review by vplasencia — CHANGES_REQUESTED** [review]
It looks great! Thanks. Just added one comment.

File: packages/circuits/semaphore.circom (hunk 1 above)
Comment (cedoor) [reply packages/circuits/semaphore.circom:28]: I agree. I've added some comments. Let me know if they're ok.

File: packages/circuits/semaphore.circom [hunk 2]
```diff
@@ -4,22 +4,63 @@ include "babyjub.circom";
 include "poseidon.circom";
 include "binary-merkle-root.circom";
 
+// The Semaphore circuit can be divided into 3 main parts.
+// The first part involves the generation of the Semaphore identity,
+// i.e. the public key and its hash, which is called the commitment
+// and is used as a public value.
+// In the second part, it is verified whether the identity commitment is part
+// of the Merkle tree, i.e. the Semaphore group. That is, a proof of membership is verified.
+// The third part covers the generation of a nullifier, i.e. the hash of the secret used to
```
Comment (vplasencia) [inline packages/circuits/semaphore.circom]: What about saying the values of the hash in order, it would be:

```
The third part covers the generation of a nullifier, i.e. the hash of the scope of the proof and the secret used to derive the public key.
```

File: packages/circuits/semaphore.circom [hunk 3]
```diff
@@ -4,22 +4,63 @@ include "babyjub.circom";
...
+// the user, or the hash of that value.
 template Semaphore(MAX_DEPTH) {
+    // Input signals.
+    // The input signals are all private except 'message' and 'scope'.
+    // The secret is the scalar generated from the EdDSA private key.
+    // Using the secret scalar instead of the private key allows this circuit
+    // to skip steps 1, 2, 3 in the generation of the public key defined here:
+    // https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5, making the circuit
+    // more efficient and simple.
+    // See the Semaphore identity package to know more about how the identity is generated.
```
Comment (vplasencia) [inline packages/circuits/semaphore.circom]: It would be nice to add the Semaphore identity package link to easily go there.

**Review by vplasencia — CHANGES_REQUESTED** [review]

File: packages/circuits/semaphore.circom [hunk 4]
```diff
@@ -4,22 +4,64 @@ include "babyjub.circom";
...
 include "binary-merkle-root.circom";
 
+// The Semaphore circuit can be divided into 3 main parts.
+// The first part involves the generation of the Semaphore identity,
+// i.e. the public key and its hash, which is called the commitment
+// and is used as a public value.
+// In the second part, it is verified whether the identity commitment is part
+// of the Merkle tree, i.e. the Semaphore group. That is, a proof of membership is verified.
+// The third part covers the generation of a nullifier, i.e. the hash of the scope of the proof
+// and the secret used to derive the public key. The nullifier is used to prevent the same
```
Comment (vplasencia) [inline packages/circuits/semaphore.circom]:
```suggestion
// and the secret used to derive the public key (secret scalar). The nullifier is used to prevent the same
```

**Review by vplasencia — CHANGES_REQUESTED** [review]

**Review by vplasencia — APPROVED** [review]

File: packages/circuits/semaphore.circom [hunk 5]
```diff
@@ -4,22 +4,64 @@ include "babyjub.circom";
 include "poseidon.circom";
 include "binary-merkle-root.circom";
 
+// The Semaphore circuit can be divided into 3 main parts.
+// The first part involves the generation of the Semaphore identity,
+// i.e. the public key and its hash, which is called the commitment
```
Comment (0xjei) [inline packages/circuits/semaphore.circom:9]: "... which is called the commitment ...". Can be confusing: public key or hash or both of them?

File: packages/circuits/semaphore.circom [hunk 6]
```diff
@@ -4,22 +4,64 @@ include "babyjub.circom";
...
+// and is used as a public value.
+// In the second part, it is verified whether the identity commitment is part
```
Comment (0xjei) [inline packages/circuits/semaphore.circom]: "In the second part, it is verified whether or not the identity commitment is part of the Semaphore group, implemented as a Merkle tree." 

I would rephrase with something like this

File: packages/circuits/semaphore.circom [hunk 7]
```diff
@@ -4,22 +4,64 @@ include "babyjub.circom";
...
+// of the Merkle tree, i.e. the Semaphore group. That is, a proof of membership is verified.
+// The third part covers the generation of a nullifier, i.e. the hash of the scope of the proof
+// and the secret used to derive the public key (secret scalar). The nullifier is used to prevent the same
```
Comment (0xjei) [inline packages/circuits/semaphore.circom:14]: Probably worth adding an high level representation:

// h(scope, secret) ...

File: packages/circuits/semaphore.circom [hunk 8]
```diff
@@ -4,22 +4,64 @@ include "babyjub.circom";
...
+// proof from being verified twice.
+// The circuit lastly includes the message, which is an arbitrary anonymous value defined by
+// the user, or the hash of that value.
 template Semaphore(MAX_DEPTH) {
+    // Input signals.
+    // The input signals are all private except 'message' and 'scope'.
+    // The secret is the scalar generated from the EdDSA private key.
+    // Using the secret scalar instead of the private key allows this circuit
+    // to skip steps 1, 2, 3 in the generation of the public key defined here:
```
Comment (0xjei) [inline packages/circuits/semaphore.circom:23]: very neat here!

**Review by 0xjei — COMMENTED** [review]
Good work, very precise. Broadly speaking, I think it will do. I took the liberty of adding some thoughts on how to improve imho but not as changes needed for the merge.

File: packages/circuits/semaphore.circom (hunk 5 above)
Comment (cedoor) [reply packages/circuits/semaphore.circom:9]: The hash, "which" is after "hash,", so it should be pretty clear

File: packages/circuits/semaphore.circom (hunk 7 above)
Comment (cedoor) [reply packages/circuits/semaphore.circom:14]: I think the high-level representation is the code itself at line `60`. These comments will not be part of TypeDoc, so they will still see the code.
